### PR TITLE
ci: update pre-commit yaml file and add as workflow in .github

### PR DESCRIPTION
**Pull request by benwandrew**
# Description

update `.pre-commit-config.yaml` and add a new workflow for running the hooks upon PR

# Type of change

- **ci**: Changes to our CI configuration files and scripts (i.e., those in the .github directory)

### CODE REVIEW

**Comment by TheLemonPig** [comment]
I am confused by why this is crashing. Do we need to commit changes after running ```pre-commit run --all-files```? (This is what I needed to do for the BMS Tutorial.

**Review by TheLemonPig — CHANGES_REQUESTED** [review]
It looks like the same issue as with the equivalent PR for ```autora-theorist-bsr```

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,28 +1,37 @@
...
         - "--max-line-length=100"
         - "--extend-ignore=E203"
+        - "--per-file-ignores=__init__.py:F401"
   - repo: https://github.com/pre-commit/mirrors-mypy
-    rev: "v0.991"
+    rev: "v1.5.1"
     hooks:
       - id: mypy
-        additional_dependencies: [types-requests]
+        additional_dependencies: [types-requests,scipy,pytest]
```
Comment (hollandjg) [inline .pre-commit-config.yaml]: @benwandrew have you tried including the additional types here?

Comment (benwandrew) [reply .pre-commit-config.yaml]: resolved by adding relevant overrides

`[[tool.mypy.overrides]]
module = ["matplotlib.*", "sklearn.*", "pandas.*", "numpy.*","scipy.*"]
ignore_missing_imports=true`

**Review by TheLemonPig — APPROVED** [review]
Looks great!

**Pull request merged**